### PR TITLE
remove .data from torch/jit

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31482 remove simple .data from torch/nn
* #31481 remove .data from torch/nn doc
* **#31480 remove .data from torch/jit**
* #31479 Remove some .data and clearly mark why we keep some

Differential Revision: [D19303244](https://our.internmc.facebook.com/intern/diff/D19303244)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 41685f1:


* **1/2** broken upstream at merge base fc3103b (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=fc3103b1162a3312e4ad74005d0b698a333c0750&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **1/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=41685f1f33bcb9b82307fd0d03c8d5e30bc97288).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4017714) caffe2_onnx_py2_gcc5_ubuntu16_04_build (1/1)
**Step:** "Set Up CI Environment After attach_workspace" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77605407) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77605407))
<details>
<summary>
<code>E: Failed to fetch https://nvidia.github.io/nvidia-container-runtime/ubuntu16.04/amd64/./nvidia-container-runtime_2.0.0+docker18.09.4-1_amd64.deb  GnuTLS recv error (-54): Error in the pull function.</code>
</summary>

```
                                                                   91% [Working]               Get:26 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 intel-microcode amd64 3.20191115.1ubuntu0.16.04.2 [2,408 kB] 
91% [26 intel-microcode 64.1 kB/2,408 kB 3%] 94% [Working]                                                      21.3 MB/s 0s                                                                                 Get:27 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 amd64-microcode amd64 3.20191021.1+really3.20180524.1~ubuntu0.16.04.2 [30.8 kB] 
94% [27 amd64-microcode 20.0 kB/30.8 kB 65%]                       21.3 MB/s 0s 94% [Working]                                                      21.3 MB/s 0s                                                                                 Get:28 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 linux-image-generic amd64 4.4.0.170.178 [2,446 B] 
94% [28 linux-image-generic 2,297 B/2,446 B 94%]                   21.3 MB/s 0s 95% [Working]                                                      21.3 MB/s 0s                                                                                 Get:29 http://archive.ubuntu.com/ubuntu xenial/universe amd64 moreutils amd64 0.57-1 [55.0 kB] 
95% [29 moreutils 2,295 B/55.0 kB 4%]                              21.3 MB/s 0s 95% [Working]                                                      21.3 MB/s 0s                                                                                 Get:30 http://archive.ubuntu.com/ubuntu xenial/main amd64 tcl8.6 amd64 8.6.5+dfsg-2 [14.2 kB] 
95% [30 tcl8.6 2,295 B/14.2 kB 16%]                                21.3 MB/s 0s 96% [Working]                                                      21.3 MB/s 0s                                                                                 Get:31 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 thermald amd64 1.5-2ubuntu4 [187 kB] 
96% [31 thermald 2,293 B/187 kB 1%]                                21.3 MB/s 0s 97% [Working]                                                      21.3 MB/s 0s                                                                                 Get:32 http://archive.ubuntu.com/ubuntu xenial/main amd64 tk8.6 amd64 8.6.5-1 [12.2 kB] 
97% [32 tk8.6 2,295 B/12.2 kB 19%]                                 21.3 MB/s 0s 97% [Working]                                                      21.3 MB/s 0s                                                                                 Get:33 http://archive.ubuntu.com/ubuntu xenial/main amd64 xterm amd64 322-1ubuntu1 [607 kB] 
97% [33 xterm 2,293 B/607 kB 0%]                                   21.3 MB/s 0s 98% [Working]                                                      21.3 MB/s 0s                                                                                 Fetched 129 MB in 7s (17.7 MB/s) 
W: --force-yes is deprecated, use one of the options starting with --allow instead. 
E: Failed to fetch https://nvidia.github.io/nvidia-container-runtime/ubuntu16.04/amd64/./nvidia-container-runtime_2.0.0+docker18.09.4-1_amd64.deb  GnuTLS recv error (-54): Error in the pull function. 
 
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing? 
```

</details>

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_6_gcc5_4_test](https://dr.pytorch.org/build-details.html?build_id=77616860)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Review by zdevito — APPROVED** [review]

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -379,7 +379,7 @@ def clone_input(a):
             return None
         elif isinstance(a, torch.Tensor):
             # TODO: figure out one liner to .clone() and set requires_grad
-            v = Variable(a.data.clone(memory_format=torch.preserve_format), requires_grad=a.requires_grad)
+            v = a.detach().clone(memory_format=torch.preserve_format).requires_grad_(a.requires_grad)
```
Comment (gchanan) [inline torch/jit/__init__.py]: for these `detach().clone(...)` calls -- why not put them in a no_grad block?  It seems generally preferable to have this order of preference:
1) no_grad
2) detach
3) .data

because you don't actually create this intermediate Tensor (`a.detach()`) that is potentially floating around.

**Review by gchanan — APPROVED** [review]

File: torch/jit/__init__.py (hunk 1 above)
Comment (albanD) [reply torch/jit/__init__.py]: The version with no_grad does not look much clearer here I think:
```python
with torch.no_grad():
  v = a.clone(memory_format=torch.preserve_format)
v.requires_grad_(a.requires_grad)
```
The one liner above is clearer about what we want to do. And the intermediate Tensor is quite cheap.

Comment (gchanan) [reply torch/jit/__init__.py]: can you not set requires_grad under a no_grad block?  That doesn't seem right.

Comment (albanD) [reply torch/jit/__init__.py]: As discussed offline, you can. But it is trickier as the user need to be aware if the operation returns a new Tensor or not. Will keep it as is for now.
